lib/compiler/src/beam_ssa_bc_size.erl

```diff
@@ -0,0 +1,596 @@
+    case get_arg_value(Arg, Bs) of
+        none ->
+            Bs#{Dst => any};
+        Val ->
+            Bs#{Dst => {match,{{bif,bit_size},[Val]},#b_literal{val=0}}}
+    end;
+calc_size_instr(#b_set{op=call,args=[Name|Args],dst=Dst}=I, Bs) ->
+    if
+        is_map_key(Name, Bs) ->
+            Result0 = [get_value(A, Bs) || A <- Args],
+            Result = [Val || Val <- Result0, Val =/= any],
+            Bs#{Dst => {call,Result}};
+        true ->
+            call_size_func(I, Bs)
+    end;
+calc_size_instr(#b_set{op=get_tl,args=[Ctx],dst=Dst}, Bs) ->
+    update_match(Dst, Ctx, #b_literal{val=1}, Bs);
+calc_size_instr(#b_set{op=is_nonempty_list,args=[Arg],dst=Dst}, Bs) ->
+    case get_arg_value(Arg, Bs) of
+        none ->
+            Bs#{Dst => any};
+        Val ->
+            NumElements = {{bif,length},[Val]},
+            Match = {match,NumElements,#b_literal{val=0}},
+            NoMatch = {nil_or_bad,Val},
+            Bs#{Dst => {'if',Arg,Match,NoMatch}}
+    end;
+calc_size_instr(#b_set{op=put_tuple,args=Args,dst=Dst}, Bs) ->
+    Bs#{Dst => {tuple,Args}};
+calc_size_instr(#b_set{op={succeeded,_},args=[Arg],dst=Dst}, Bs) ->
+    Bs#{Dst => {succeeded,Arg}};
+calc_size_instr(#b_set{dst=Dst}, Bs) ->
+    Bs#{Dst => any}.
+
+update_writable(Dst, Writable, Expr, Bs) ->
+    {writable,#b_literal{val=0}} = get_value(Writable, Bs),
+    Bs#{Dst => {writable,Expr}}.
+
+update_match(Dst, Ctx, Increment, Bs) ->
+    case get_value(Ctx, Bs) of
+        {match,NumElements,Offset0} ->
+            Offset = {{bif,'+'},[Offset0,Increment]},
+            Bs#{Dst => {match,NumElements,Offset}};
+        _ ->
+            Bs#{Dst => any}
+    end.
+
+get_arg_value(#b_literal{}=Lit, _Bs) ->
+    Lit;
+get_arg_value(Name, Bs) ->
+    case Bs of
+        #{Name := {arg,Val}} -> Val;
+        #{} -> none
+    end.
+
+get_value(Name, Bs) ->
+    case Bs of
+        #{Name := Value} -> Value;
+        #{} -> Name
+    end.
+
+join_bs(LHS, RHS) ->
+    if
+        map_size(LHS) < map_size(RHS) ->
+            join_bs_1(maps:keys(LHS), RHS, LHS);
+        true ->
+            join_bs_1(maps:keys(RHS), LHS, RHS)
+    end.
+
+%% Joins two maps of bindings, keeping the variables that are common to both maps.
+join_bs_1([V | Vs], Bigger, Smaller) ->
+    case {Bigger, Smaller} of
+        {#{V := Same},#{V := Same}} ->
+            join_bs_1(Vs, Bigger, Smaller);
+        {#{V := _LHS},#{V := _RHS}} ->
+            join_bs_1(Vs, Bigger, Smaller#{V := any});
+        {#{}, #{V := _}} ->
+            join_bs_1(Vs, Bigger, maps:remove(V, Smaller))
+    end;
+join_bs_1([], _Bigger, Smaller) -> Smaller.
+
+%%%
+%%% Turn the result of the traversal of the SSA code into an expression tree.
+%%%
+
+make_expr_tree({{call,Alloc0},GenAnno}) ->
+    {Alloc1,Annos} = make_expr_tree_list(Alloc0, none, none, [GenAnno]),
+    Alloc2 = opt_expr(Alloc1),
+    Alloc = case divisible_by_eight(Alloc2) of
```

Comment:
I think you should add a comment or break this out into its own function. At a first glance it's not obvious we're rounding to the next byte.